templates/test/ci/prow-clusterclass-ci-default/base.yaml

```diff
@@ -0,0 +1,264 @@
+spec:
+  template:
+    spec:
+      kubeadmConfigSpec:
+        clusterConfiguration:
+          apiServer:
+            extraArgs:
+              cloud-config: /etc/kubernetes/azure.json
+              cloud-provider: azure
+            extraVolumes:
+              - hostPath: /etc/kubernetes/azure.json
+                mountPath: /etc/kubernetes/azure.json
+                name: cloud-config
+                readOnly: true
+            timeoutForControlPlane: 20m
+          controllerManager:
+            extraArgs:
+              allocate-node-cidrs: "false"
+              cloud-config: /etc/kubernetes/azure.json
+              cloud-provider: azure
+            extraVolumes:
+              - hostPath: /etc/kubernetes/azure.json
+                mountPath: /etc/kubernetes/azure.json
+                name: cloud-config
+                readOnly: true
+          etcd:
+            local:
+              dataDir: /var/lib/etcddisk/etcd
+              extraArgs:
+                quota-backend-bytes: "8589934592"
+        diskSetup:
+          filesystems:
+            - device: /dev/disk/azure/scsi1/lun0
+              extraOpts:
+                - -E
+                - lazy_itable_init=1,lazy_journal_init=1
+              filesystem: ext4
+              label: etcd_disk
+            - device: ephemeral0.1
+              filesystem: ext4
+              label: ephemeral0
+              replaceFS: ntfs
+          partitions:
+            - device: /dev/disk/azure/scsi1/lun0
+              layout: true
+              overwrite: false
+              tableType: gpt
+        files:
+          - contentFrom:
+              secret:
+                key: control-plane-azure.json
+                name: replace_me
+            owner: root:root
+            path: /etc/kubernetes/azure.json
+            permissions: "0644"
+        initConfiguration:
+          nodeRegistration:
+            kubeletExtraArgs:
+              azure-container-registry-config: /etc/kubernetes/azure.json
+              cloud-config: /etc/kubernetes/azure.json
+              cloud-provider: azure
+            name: '{{ ds.meta_data["local_hostname"] }}'
+        joinConfiguration:
+          nodeRegistration:
+            kubeletExtraArgs:
+              azure-container-registry-config: /etc/kubernetes/azure.json
+              cloud-config: /etc/kubernetes/azure.json
+              cloud-provider: azure
+            name: '{{ ds.meta_data["local_hostname"] }}'
+        mounts:
+          - - LABEL=etcd_disk
+            - /var/lib/etcddisk
+        postKubeadmCommands: [ ]
+        preKubeadmCommands: [ ]
+---
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: AzureClusterTemplate
+metadata:
+  name: ci-default-azure-cluster
+spec:
+  template:
+    spec:
+      additionalTags:
+        replace_me_key: replace_me_val
+      identityRef:
+        apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+        kind: AzureClusterIdentity
+      location: replace_me
```

Comment:
just curious, is "replace_me" an arbitrary string or is it some agreed upon contract/pattern for required variables in clusterclass?

Reply:
It's unfortunately just a workaround for fields that have to be set for the validating webhook and are then later overwritten per cluster.

I think this workaround kind of becomes a pattern right now, but there is no standardization right now.

Not sure how we can improve this, maybe it's an option that these fields should be optional when the resource is reference in a ClusterClass. But not sure how feasible this is.

(as far as I'm aware CAPA is using a similar but not the same string)

Reply:
This is just an arbitrary string to indicate that the value will be coming from a variable. There is no agreed upon contract but I vaguely remember this convention being used elsewhere.

Reply:
CAPA is using [REPLACEME](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/318413aaded7f7aece86f315af7058bd956ec488/templates/cluster-template-multitenancy-clusterclass.yaml#L189). I don't particularly have a preference but I vaguely remember using `replace_me` being used in other places where a similar mechanism was involved.

Reply:
I'm not sure, we probably have some guidance around cluster-specific fields and that they shouldn't be added to cluster templates (which was in general nicely done by CAPZ). But I"m not sure if that falls under the "cluster-specific" field category.

I think it makes sense to open an issue in CAPI and see what we have / need.

We can discuss documentation improvements or potential implementation changes to mitigate this issue (e.g. maybe some fields shouldn't be mandatory on a cluster template / machine template when it's used as template in a ClusterClass).